### PR TITLE
Add basic option to compute second derivative

### PR DESCRIPTION
**Pull request by timmens**
This will be the first PR in a series to add basic and advanced features regarding the second derivative to estimagic. The PR-series is related to issue #232.

In this PR we tackle the first bullet point in #232, that is, we implement the numerical second derivative with parallel function evaluations.

### CODE REVIEW

**Comment by timmens** [comment]
Things that remain to be done:

- [ ] Write tests
- [ ] Find bug which produces wrong results on off-diagonal
- [ ] Capture two and cross step evaluations in data frame

**Comment by timmens** [comment]
Problem: statsmodels implementation is much more accurate which seems to be due to the fact that they are using more complicated steps. Before closing this PR we need to check if we want to adopt this or stick to the simpler version and sacrifice accuracy.

File: src/estimagic/differentiation/derivatives.py [hunk 1]
```diff
@@ -244,11 +244,376 @@ def first_derivative(
...
+    error_handling="continue",
+    batch_evaluator="joblib",
+    return_func_value=False,
+    return_info=True,
+    key=None,
+):
+    """Evaluate second derivative of func at params according to method and step options
+
+    Internally, the function is converted such that it maps from a 1d array to a 1d
+    array. Then the Hessian of that function is calculated. The resulting derivative
```
Comment (janosg) [inline src/estimagic/differentiation/derivatives.py:277]:
```suggestion
    array. Then the Hessians of that function are calculated. The resulting derivative
```

**Review by janosg — APPROVED** [review]
Well done! Very nice PR! Thanks a lot!

**Pull request merged**